[최용석] STEP1 PR 요청드립니다.

badaelephant opened the pull request:
안녕하세요 장민제 리뷰어님. 

이번에 인수테스트 기반 TDD를 같이 하게 되었네요 ㅎㅎ 

잘 부탁드립니다.

Commit:
docs: step1 리드미 작성

Commit:
refactor: 도메인으로 옮길 로직 찾기, 도메인 단위테스트 목록 작성하기

Commit:
test: 거리, 구간 단위 테스트 작성

Commit:
feat: 거리와 구간 기능 추가 - 테스트 통과'

Commit:
test: 노선 테스트 생성, 거리테스트 수정, 노선_레포지토리 테스트 생성

Commit:
test: 구간_목록 테스트 생성

Commit:
test: 테스트의 예외 IllegalStateException 수정

Commit:
test: 구간 및 구간목록 테스트수정
꼭 필요한 부분만 getter 사용하도록

Commit:
feat: 구간 및 구간목록 기능 생성

Commit:
refactor: 노선 기능 수정 - getSections 안하도록
노선테스트 일부 수정

Commit:
refactor: 노선 서비스 수정

Commit:
refactor: 노선 컨트롤러 수정

Commit:
refactor: 에러 메시지 따로 빼기

Commit:
인수조건 통합테스트 추가

Commit:
fix: 실패하는 테스트 부분 수정

Commit:
docs: step1 리드미 수정부분 체크

minzzang commented on README.md:
```diff
@@ -1,54 +1,39 @@
+### 요구사항
 
-<br>
+- [x] LineService 리팩터링 
```
요구사항 정리 👍

minzzang commented on src/main/java/nextstep/subway/ErrMsg.java:
```diff
@@ -0,0 +1,28 @@
+package nextstep.subway;
+
+public class ErrMsg {
```
예외 메시지 분리 잘해주셨네요! 객체지향 생활 체조에도 나와 있듯이 축약어를 지양해보면 어떨까요~?

badaelephant replied:
축약어는 안쓰는 것이 좋군요. 확인하고 수정하였습니다.

minzzang commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -28,155 +27,65 @@ public LineService(LineRepository lineRepository, StationService stationService)
+    private Line createLineFromRequest(LineRequest request){
+        Station upStation = stationService.findById(request.getUpStationId());
+        Station downStation = stationService.findById(request.getDownStationId());
+        Line line = request.toLine();
```
request에서 Line으로 변환할 때 완전한 상태로 변환해보면 어떨까요~?

badaelephant replied:
아 정말 그렇네요, lineRequest에서 목적에 따라(update name or color / 정상적인 line 객체의 생성) 등에 따라 불완전(구간 없음) / 완전(구간 포함) 노선 객체를 사용하는 것 같아서. lineRequest에서 2가지 변환 메소드로 나누었습니다. 혹시 이부분 문제 없는지 확인 한번 부탁드립니다.

minzzang commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -28,155 +27,65 @@ public LineService(LineRepository lineRepository, StationService stationService)
+        line.addSection(new Section(line, upStation, downStation, request.getDistance()));
+        return line;
+    }
+    private LineResponse createLineResponse(Line line){
```
이 부분도 LineResponse를 생성할 때 line을 넘겨주면 서비스 레이어에서 변환 로직을 없앨 수 있을 거 같네요! 😄

badaelephant replied:
아 정말 그렇군요. 서비스 레이어는 아무래도 최대한 서비스 그 자체의 동작이 이루어지는 것으로 역할을 한정시키는게 좋을것 같습니다. response에 대한 정보 제공(위 경우에서는 getOrderedStations)의 부분은 LineResponse에서 돌아가는 것이 좋겠네요

minzzang commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,45 @@
+import javax.persistence.Embeddable;
+
+@Embeddable
+public class Distance {
```
Distance 래핑 👍

minzzang commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -28,155 +27,65 @@ public LineService(LineRepository lineRepository, StationService stationService)
-
+    @Transactional
     public void updateLine(Long id, LineRequest lineUpdateRequest) {
         Line persistLine = lineRepository.findById(id).orElseThrow(RuntimeException::new);
```
findLineById메소드를 사용해도 좋을 것 같아요! 😄

minzzang commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -28,155 +27,65 @@ public LineService(LineRepository lineRepository, StationService stationService)
-
-    public void addLineStation(Long lineId, SectionRequest request) {
+    @Transactional
+    public void addLineStation(Long lineId, SectionRequest request) throws NotFoundException {
```
RuntimeException이 아니라 Exception을 사용하신 이유가 있을까요~?

badaelephant replied:
아 제가 Exception 핸들러에 대해서 확실히 신경 못쓰고 있었는데 잡아주셔서 감사합니다. 포괄적인 Exception 클래스를 상속받은 NotFoundException 보다 RuntimeException을 상속받은 EntityNotFoundException 커스텀 클래스를 만들었습니다.

minzzang commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -28,155 +27,65 @@ public LineService(LineRepository lineRepository, StationService stationService)
-        } else {
-            throw new RuntimeException();
-        }
+        Section section = new Section(line, upStation, downStation, request.getDistance());
```
도메인 로직을 잘 분리해주셨네요 💯

minzzang commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -50,7 +51,18 @@ public String getColor() {
+    }
+
+    public List<Station> getOrderedStations() {
+        return sections.getOrderedStations();
```
Unmodifiable Collection을 활용해보면 어떨까요~?

badaelephant replied:
제가 이부분은 정확하게 이해가 안되었는데, 
Sections에서 getOrderedStation을 할때 unmodifiableList를 반환해주고 잇습니다. 혹시 이 경우에 returntype을 List<Station> 이 아닌 다른 것으로 바꿔줘야 할까요?

minzzang replied:
아 제가 잘못 보고 피드백을 드렸네요!! 🙏 return은 List로 두셔도 됩니다~

minzzang commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,189 @@
+        }
+    }
+
+    private boolean addUpMatchingSection(Section section) {
```
두 가지 책임이 있는것 같은데 분리해보면 어떨까요~?

badaelephant replied:
확인하고 분리하였습니다. ㅎㅎ 감사합니다.^^ 역할 분리! 명심해야겠네요

minzzang commented on src/test/java/nextstep/subway/line/unit/SectionsTest.java:
```diff
@@ -0,0 +1,178 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class SectionsTest {
```
도메인 단위 테스트 👍

minzzang commented on src/test/java/nextstep/subway/line/acceptance/LineSectionAcceptanceTest.java:
```diff
@@ -42,6 +42,50 @@ public void setUp() {
+     */
+    @DisplayName("지하철 구간 관련 기능 통합 테스트")
+    @Test
+    void 통합테스트() {
```
인수 테스트 잘 작성해주셨네요! 👍 dynamic test를 적용해봐도 좋을 것 같아요!

minzzang reviewed (changes requested):
안녕하세요 용석님~!

인수테스트 기반 TDD 1단계 미션 잘 작성해주셨네요! :+1:

소소한 코멘트 남겼는데 확인 부탁드립니다. :pray:

이번 주 화이팅입니다. :muscle:

Commit:
refactor: 피드백 반영
1. 축약어 지양 ErrMsg-> ErrorMessage
2. LineRequest에서 Line만들때 완전하게 넘겨주기
3. LineResponse 생성시 Line을 넘겨주어 서비스 레이어에서 변환로직을 제거

Commit:
refactor: 피드백 반영
Sections에서 addSectionInMiddle 에 대하여 책임을 분리

Commit:
refactor: 피드백 반영
Dynamic Test 적용

Commit:
refactor: 정적 팩토리 메소드 Line, Section 적용

Commit:
refactor: Exception->RuntimeException으로 수정(커스텀 Exception 사용), ErrorResponse 추가.

minzzang commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -150,15 +147,18 @@ private Station getNextStation(Station station) {
 
-    private void validateDeleteCondition() {
+    private void validateDeleteCondition(Station station) {
+        if(!isExists(station)){
```
전체적으로 컨벤션이 안맞는거 같은데 우테코 스타일 적용해보는건 어떨까요~?
https://programmer-ririhan.tistory.com/338

minzzang commented on src/main/java/nextstep/subway/line/dto/LineRequest.java:
```diff
@@ -40,7 +41,10 @@ public int getDistance() {
+    public Line toLine(Station upStation, Station downStation) {
+        return Line.of(name, color, upStation, downStation, distance);
+    }
+    public Line toUpdateLineInfo(){
```
이런식으로 해도 괜찮을 거 같긴합니다! 😄  다만, create와 update를 할 때 같은 dto를 사용하면 어떤 문제가 생길 수 있을까요~?

minzzang commented on src/test/java/nextstep/subway/line/acceptance/LineSectionAcceptanceTest.java:
```diff
@@ -42,6 +49,61 @@ public void setUp() {
+                }),
+                dynamicTest("노선에 등록된 역 목록을 조회하여 역을 순차적으로 조회한다.", () -> {
+
+                    // when 지하철 노선에 등록된 역 목록 조회 요청
```
dynamic Test 잘 적용하셨네요!! 👍  람다 부분을 메소드로 추출하면 가독성이 더 좋아질 거 같아요!

minzzang reviewed (approved):
안녕하세요 용석님~!

피드백 잘 반영해주셨네요! :+1:

소소한 코멘트 남겼는데 확인 부탁드립니다. :pray:

그럼 다음 미션도 화이팅입니다. :muscle:
